README.md
# gender_picker 1.1.0

An awesome gender picker widget comes with latest feature to support in your project design.

The goal of this project is to provide an ultimate widget for selecting the gender.

The source code is 100% Dart, and everything resides in the /lib folder.

![Screenshot_20200314_120050_com mateen example](https://user-images.githubusercontent.com/24449076/76674642-b1ece100-65ec-11ea-80a6-3327ac72b4d3.jpg)

## Installing
Add this to your package's pubspec.yaml file:

dependencies:
``` gender_picker: ^1.1.0 ```


## Sample Usage
```
import 'package:gender_picker/source/enums.dart';
import 'package:gender_picker/source/gender_picker.dart';
GenderPickerWithImage(
              showOtherGender: true,
              verticalAlignedText: false,
              selectedGender: Gender.Male,
              selectedGenderTextStyle: TextStyle(
                  color: Color(0xFF8b32a8), fontWeight: FontWeight.bold),
              unSelectedGenderTextStyle: TextStyle(
                  color: Colors.white, fontWeight: FontWeight.normal),
              onChanged: (Gender gender) {
                print(gender);
              },
              equallyAligned: true,
              animationDuration: Duration(milliseconds: 300),
              isCircular: true,
              // default : true,
              opacityOfGradient: 0.4,
              padding: const EdgeInsets.all(3),
              size: 50, //default : 40
            )
```

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [1.1.0] - Release April 15, 2021.

* Upgrade to null safety

## [1.0.4] - Release July 20, 2020.

* Update gender options

## [1.0.3] - Release July 20, 2020.

* Change animation

## [1.0.2] - Release March 14, 2020.

* Updated vertical view alignment

## [1.0.1] - Release March 14, 2020.

* Changes in classes for effecient view

## [1.0.0] - Release March 14, 2020.

* Updated with latest bugfree version

## [0.0.3] - Release March 14, 2020.

* Updated with immutation


## [0.0.2] - Release March 14, 2020.

* Updated with example to use properly

## [0.0.1] - Release March 13, 2020.

* Released full changelog with readme.

test/gender_picker_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:gender_picker/gender_picker.dart';


lib/source/gender_picker.dart
import 'package:flutter/material.dart';
import 'package:gender_picker/source/enums.dart';

// ignore: must_be_immutable
class GenderPickerWithImage extends StatefulWidget {
  // optional: selectedGender can be null: if null, by default nothing will be selected
  Gender? selectedGender;

  // to align title under picture, set true
  final bool verticalAlignedText;

  // gradient color of highlighting text
  final LinearGradient linearGradient;

  // to show images to respectively
  final ImageProvider maleImage;
  final ImageProvider femaleImage;
  final ImageProvider otherGenderImage;

  // on selection of gender changed
  final ValueChanged<Gender?>? onChanged;

  // Title of gender
  final String maleText;
  final String femaleText;
  final String otherGenderText;

  // alignment and difference between each item
  final bool equallyAligned;

  // size of gender icon
  final double size;

  // selection required circular or in square
  final bool isCircular;

  // padding of widget
  final EdgeInsetsGeometry padding;

  // time taken to show full opacity
  final Duration animationDuration;

  // required opacity of gradient color
  final double opacityOfGradient;

  // to define color of selected and unselected text
  final TextStyle selectedGenderTextStyle;
  final TextStyle unSelectedGenderTextStyle;

  // to show other genders
  final bool showOtherGender;

  GenderPickerWithImage({
    Key? key,
    // by default selected color of selected gender text
    this.selectedGenderTextStyle = const TextStyle(
      fontSize: 19,
      color: Colors.redAccent,
      fontWeight: FontWeight.w600,
    ),
    // by default un-selected color of selected gender text
    this.unSelectedGenderTextStyle = const TextStyle(
      fontSize: 19,
      color: Colors.black,
      fontWeight: FontWeight.w600,
    ),

    // bedefault verticallyaligned text is false to show in same row
    this.verticalAlignedText = false,

    // animation loading time by default set to [500 milliseconds]
    this.animationDuration = const Duration(milliseconds: 500),

    // by default gradient opacity set to 0.6
    this.opacityOfGradient = 0.6,
    // padding set to 10 from all side, you can also define yours
    this.padding = const EdgeInsets.all(10),

    // selection is in circular and can be in square as you needed
    this.isCircular = true,

    // default size of ImageIcon
    this.size = 40.0,

    // if alignment is true then everything will be in center
    this.equallyAligned = true,

    // to show selected Gender
    this.selectedGender = Gender.Male,

    // OnChanged is mandatory to implement, so ca change state of widget
    required this.onChanged,

    // by default gradient color
    this.linearGradient = const LinearGradient(
        colors: [Color(0xFF3271a8), Color(0xFF8b32a8)],
        tileMode: TileMode.clamp,
        begin: Alignment.bottomLeft,
        end: Alignment.topRight,
        stops: [0.0, 1.0]),
    this.showOtherGender = false,

    // Images asset can be network images,
    this.maleImage =
        const AssetImage("assets/images/male.png", package: 'gender_picker'),
    this.femaleImage =
        const AssetImage("assets/images/female.png", package: 'gender_picker'),
    this.otherGenderImage =
        const AssetImage("assets/images/other.png", package: 'gender_picker'),

    // by default text to show
    this.maleText = "Male",
    this.femaleText = "Female",
    this.otherGenderText = "Other",
  }) : super(key: key);

  @override
  _GenderPickerWithImageState createState() => _GenderPickerWithImageState();
}

class _GenderPickerWithImageState extends State<GenderPickerWithImage> {
  @override
  Widget build(BuildContext context) {
    return widget.verticalAlignedText
        ? getVerticalAlignedView()
        : getHorizontalAlignedView();
  }

  Widget getVerticalAlignedView() {
    return FittedBox(
      fit: BoxFit.fitWidth,
      child: Padding(
        padding: widget.padding,
        child: Container(
          width: MediaQuery.of(context).size.width,
          child: Row(
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: widget.equallyAligned
                ? MainAxisAlignment.spaceEvenly
                : MainAxisAlignment.center,
            children: <Widget>[
              GestureDetector(
                onTap: () {
                  setState(() {
                    widget.selectedGender = Gender.Male;
                    if (widget.onChanged != null) {
                      widget.onChanged!(widget.selectedGender);
                    } else {
                      return;
                    }
                  });
                },
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Stack(
                      fit: StackFit.loose,
                      children: <Widget>[
                        Container(
                          height: widget.size,
                          width: widget.size,
                          margin: const EdgeInsets.only(top: 20, bottom: 10),
                          decoration: BoxDecoration(
                            shape: widget.isCircular
                                ? BoxShape.circle
                                : BoxShape.rectangle,
                            image: DecorationImage(
                              image: widget.maleImage,
                              fit: BoxFit.cover,
                            ),
                          ),
                          child: AnimatedOpacity(
                            opacity: widget.selectedGender == Gender.Male
                                ? widget.opacityOfGradient
                                : 0.0,
                            duration: widget.animationDuration,
                            child: widget.selectedGender == Gender.Male
                                ? Container(
                                    decoration: BoxDecoration(
                                        gradient: widget.linearGradient,
                                        shape: widget.isCircular
                                            ? BoxShape.circle
                                            : BoxShape.rectangle),
                                  )
                                : null,
                          ),
                        ),
                      ],
                    ),
                    Text(widget.maleText,
                        style: widget.selectedGender == Gender.Male
                            ? widget.selectedGenderTextStyle
                            : widget.unSelectedGenderTextStyle)
                  ],
                ),
              ),
              GestureDetector(
                onTap: () {
                  setState(() {
                    widget.selectedGender = Gender.Female;
                    if (widget.onChanged != null) {
                      widget.onChanged!(widget.selectedGender);
                    } else {
                      return;
                    }
                  });
                },
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Stack(
                      fit: StackFit.loose,
                      children: <Widget>[
                        Container(
                          height: widget.size,
                          width: widget.size,
                          margin: const EdgeInsets.only(top: 20, bottom: 10),
                          decoration: BoxDecoration(
                              shape: widget.isCircular
                                  ? BoxShape.circle
                                  : BoxShape.rectangle,
                              image: DecorationImage(
                                  image: widget.femaleImage,
                                  alignment: Alignment.center,
                                  fit: BoxFit.cover)),
                          child: AnimatedOpacity(
                            opacity: widget.selectedGender == Gender.Female
                                ? widget.opacityOfGradient
                                : 0.0,
                            duration: widget.animationDuration,
                            child: widget.selectedGender == Gender.Female
                                ? Container(
                                    decoration: BoxDecoration(
                                        gradient: widget.linearGradient,
                                        shape: widget.isCircular
                                            ? BoxShape.circle
                                            : BoxShape.rectangle),
                                  )
                                : null,
                          ),
                        ),
                      ],
                    ),
                    Text(widget.femaleText,
                        style: widget.selectedGender == Gender.Female
                            ? widget.selectedGenderTextStyle
                            : widget.unSelectedGenderTextStyle)
                  ],
                ),
              ),
              widget.showOtherGender
                  ? GestureDetector(
                      onTap: () {
                        setState(() {
                          widget.selectedGender = Gender.Others;
                          if (widget.onChanged != null) {
                            widget.onChanged!(widget.selectedGender);
                          } else {
                            return;
                          }
                        });
                      },
                      child: Column(
                        mainAxisSize: MainAxisSize.min,
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: <Widget>[
                          Stack(
                            fit: StackFit.loose,
                            children: <Widget>[
                              Container(
                                height: widget.size,
                                width: widget.size,
                                margin:
                                    const EdgeInsets.only(top: 20, bottom: 10),
                                decoration: BoxDecoration(
                                    shape: widget.isCircular
                                        ? BoxShape.circle
                                        : BoxShape.rectangle,
                                    image: DecorationImage(
                                        image: widget.otherGenderImage,
                                        alignment: Alignment.center,
                                        fit: BoxFit.cover)),
                                child: AnimatedOpacity(
                                  opacity:
                                      widget.selectedGender == Gender.Others
                                          ? widget.opacityOfGradient
                                          : 0.0,
                                  duration: widget.animationDuration,
                                  child: widget.selectedGender == Gender.Others
                                      ? Container(
                                          decoration: BoxDecoration(
                                              gradient: widget.linearGradient,
                                              shape: widget.isCircular
                                                  ? BoxShape.circle
                                                  : BoxShape.rectangle),
                                        )
                                      : null,
                                ),
                              ),
                            ],
                          ),
                          Text(widget.otherGenderText,
                              style: widget.selectedGender == Gender.Others
                                  ? widget.selectedGenderTextStyle
                                  : widget.unSelectedGenderTextStyle)
                        ],
                      ),
                    )
                  : const SizedBox(),
            ],
          ),
        ),
      ),
    );
  }

  Widget getHorizontalAlignedView() {
    return FittedBox(
      fit: BoxFit.fitWidth,
      child: Padding(
        padding: widget.padding,
        child: Container(
          width: MediaQuery.of(context).size.width,
          child: Row(
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: widget.equallyAligned
                ? MainAxisAlignment.spaceEvenly
                : MainAxisAlignment.center,
            children: <Widget>[
              GestureDetector(
                onTap: () {
                  setState(() {
                    widget.selectedGender = Gender.Male;
                    if (widget.onChanged != null) {
                      widget.onChanged!(widget.selectedGender);
                    } else {
                      return;
                    }
                  });
                },
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Stack(
                      fit: StackFit.loose,
                      children: <Widget>[
                        Container(
                          height: widget.size,
                          width: widget.size,
                          margin: const EdgeInsets.only(
                              top: 0, bottom: 0, right: 5.0),
                          decoration: BoxDecoration(
                            shape: widget.isCircular
                                ? BoxShape.circle
                                : BoxShape.rectangle,
                            image: DecorationImage(
                              image: widget.maleImage,
                              fit: BoxFit.cover,
                            ),
                          ),
                          child: AnimatedOpacity(
                            opacity: widget.selectedGender == Gender.Male
                                ? widget.opacityOfGradient
                                : 0.0,
                            duration: widget.animationDuration,
                            child: widget.selectedGender == Gender.Male
                                ? Container(
                                    decoration: BoxDecoration(
                                        gradient: widget.linearGradient,
                                        shape: widget.isCircular
                                            ? BoxShape.circle
                                            : BoxShape.rectangle),
                                  )
                                : null,
                          ),
                        ),
                      ],
                    ),
                    Text(widget.maleText,
                        style: widget.selectedGender == Gender.Male
                            ? widget.selectedGenderTextStyle
                            : widget.unSelectedGenderTextStyle)
                  ],
                ),
              ),
              GestureDetector(
                onTap: () {
                  setState(() {
                    widget.selectedGender = Gender.Female;
                    if (widget.onChanged != null) {
                      widget.onChanged!(widget.selectedGender);
                    } else {
                      return;
                    }
                  });
                },
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Stack(
                      fit: StackFit.loose,
                      children: <Widget>[
                        Container(
                          height: widget.size,
                          width: widget.size,
                          margin: const EdgeInsets.only(
                              top: 0, bottom: 0, right: 5.0),
                          decoration: BoxDecoration(
                              shape: widget.isCircular
                                  ? BoxShape.circle
                                  : BoxShape.rectangle,
                              image: DecorationImage(
                                  image: widget.femaleImage,
                                  alignment: Alignment.center,
                                  fit: BoxFit.cover)),
                          child: AnimatedOpacity(
                            opacity: widget.selectedGender == Gender.Female
                                ? widget.opacityOfGradient
                                : 0.0,
                            duration: widget.animationDuration,
                            child: widget.selectedGender == Gender.Female
                                ? Container(
                                    decoration: BoxDecoration(
                                        gradient: widget.linearGradient,
                                        shape: widget.isCircular
                                            ? BoxShape.circle
                                            : BoxShape.rectangle),
                                  )
                                : null,
                          ),
                        ),
                      ],
                    ),
                    Text(widget.femaleText,
                        style: widget.selectedGender == Gender.Female
                            ? widget.selectedGenderTextStyle
                            : widget.unSelectedGenderTextStyle)
                  ],
                ),
              ),
              widget.showOtherGender
                  ? GestureDetector(
                      onTap: () {
                        setState(() {
                          widget.selectedGender = Gender.Others;
                          if (widget.onChanged != null) {
                            widget.onChanged!(widget.selectedGender);
                          } else {
                            return;
                          }
                        });
                      },
                      child: Row(
                        mainAxisSize: MainAxisSize.min,
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: <Widget>[
                          Stack(
                            fit: StackFit.loose,
                            children: <Widget>[
                              Container(
                                height: widget.size,
                                width: widget.size,
                                margin: const EdgeInsets.only(
                                    top: 0, bottom: 0, right: 5.0),
                                decoration: BoxDecoration(
                                    shape: widget.isCircular
                                        ? BoxShape.circle
                                        : BoxShape.rectangle,
                                    image: DecorationImage(
                                        image: widget.otherGenderImage,
                                        alignment: Alignment.center,
                                        fit: BoxFit.cover)),
                                child: AnimatedOpacity(
                                  opacity:
                                      widget.selectedGender == Gender.Others
                                          ? widget.opacityOfGradient
                                          : 0.0,
                                  duration: widget.animationDuration,
                                  child: widget.selectedGender == Gender.Others
                                      ? Container(
                                          decoration: BoxDecoration(
                                              gradient: widget.linearGradient,
                                              shape: widget.isCircular
                                                  ? BoxShape.circle
                                                  : BoxShape.rectangle),
                                        )
                                      : null,
                                ),
                              ),
                            ],
                          ),
                          Text(widget.otherGenderText,
                              style: widget.selectedGender == Gender.Others
                                  ? widget.selectedGenderTextStyle
                                  : widget.unSelectedGenderTextStyle)
                        ],
                      ),
                    )
                  : Container(),
            ],
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:gender_picker/source/enums.dart';
import 'package:gender_picker/source/gender_picker.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Gender Picker',
      theme: ThemeData(
        primarySwatch: Colors.brown,
      ),
      home: MyHomePage(title: 'Gender Picker'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title = "Gender Picker"}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Container(
              child: Text(
                "Horizontal Aligned",
                style: TextStyle(fontWeight: FontWeight.bold, fontSize: 20.0),
              ),
            ),
            getWidget(false, false),
            Divider(),

            Container(
              child: Text(
                "Vertical Aligned",
                style: TextStyle(fontWeight: FontWeight.bold, fontSize: 20.0),
              ),
            ),
            getWidget(false, true),
            Divider(thickness: 8.0,),
            Container(
              child: Text(
                "Optional Other Gender\n",
                style: TextStyle(fontWeight: FontWeight.normal, fontSize: 16.0),
              ),
            ),


            // If you want to show 3rd optional gender
            Container(
              child: Text(
                "Horizontal Aligned",
                style: TextStyle(fontWeight: FontWeight.bold, fontSize: 20.0),
              ),
            ),
            getWidget(true, false),
            Divider(),

            Container(
              child: Text(
                "Vertical Aligned",
                style: TextStyle(fontWeight: FontWeight.bold, fontSize: 20.0),
              ),
            ),
            getWidget(true, true),
            Divider(),

          ],
        ),
      ),
    );
  }
  
  Widget getWidget(bool showOtherGender, bool alignVertical) {
    return Container(
      margin: EdgeInsets.symmetric(vertical: 40),
      alignment: Alignment.center,
      child: GenderPickerWithImage(
        showOtherGender: showOtherGender,
        verticalAlignedText: alignVertical,
        
        // to show what's selected on app opens, but by default it's Male
        selectedGender: Gender.Male,
        selectedGenderTextStyle: TextStyle(
            color: Color(0xFF8b32a8), fontWeight: FontWeight.bold),
        unSelectedGenderTextStyle: TextStyle(
            color: Colors.black, fontWeight: FontWeight.normal),
        onChanged: (Gender? gender) {
          print(gender);
        },
        //Alignment between icons
        equallyAligned: true,
        
        animationDuration: Duration(milliseconds: 300),
        isCircular: true,
        // default : true,
        opacityOfGradient: 0.4,
        padding: const EdgeInsets.all(3),
        size: 50, //default : 40
      ),
    );
  }
}
